src/scene_graph/camera_node.cpp

```diff
@@ -17,11 +18,101 @@ namespace wr
 
-		m_view = DirectX::XMMatrixLookAtRH(pos, DirectX::XMVectorAdd(forward, pos), up);
-		m_projection = DirectX::XMMatrixPerspectiveFovRH(m_fov, m_aspect_ratio, 1, 125);
+		m_view = DirectX::XMMatrixLookAtRH(pos, DirectX::XMVectorAdd(pos, forward), up);
+		m_projection = DirectX::XMMatrixPerspectiveFovRH(m_fov, m_aspect_ratio, m_frustum_near, m_frustum_far);
+		m_view_projection = DirectX::XMMatrixMultiply(m_view, m_projection);
 		m_inverse_projection = DirectX::XMMatrixInverse(nullptr, m_projection);
 
+		CalculatePlanes();
+
 		SignalUpdate(frame_idx);
 	}
 
+	//Frustum culling code;
+	//optimized and refactored version of
+	//https://www.braynzarsoft.net/viewtutorial/q16390-34-aabb-cpu-side-frustum-culling
+
+	void CameraNode::CalculatePlanes()
+	{
+		//Left plane
+
+		m_planes[0] = DirectX::XMPlaneNormalize({
+			m_view_projection.r[0].m128_f32[3] + *m_view_projection.r[0].m128_f32,
+			m_view_projection.r[1].m128_f32[3] + *m_view_projection.r[1].m128_f32,
+			m_view_projection.r[2].m128_f32[3] + *m_view_projection.r[2].m128_f32,
+			m_view_projection.r[3].m128_f32[3] + *m_view_projection.r[3].m128_f32
+		});
+
+		//Right plane
+
+		m_planes[1] = DirectX::XMPlaneNormalize({
+			m_view_projection.r[0].m128_f32[3] - *m_view_projection.r[0].m128_f32,
+			m_view_projection.r[1].m128_f32[3] - *m_view_projection.r[1].m128_f32,
+			m_view_projection.r[2].m128_f32[3] - *m_view_projection.r[2].m128_f32,
+			m_view_projection.r[3].m128_f32[3] - *m_view_projection.r[3].m128_f32
+		});
+
+		//Top plane
+
+		m_planes[2] = DirectX::XMPlaneNormalize({
+			m_view_projection.r[0].m128_f32[3] - m_view_projection.r[0].m128_f32[1],
+			m_view_projection.r[1].m128_f32[3] - m_view_projection.r[1].m128_f32[1],
+			m_view_projection.r[2].m128_f32[3] - m_view_projection.r[2].m128_f32[1],
+			m_view_projection.r[3].m128_f32[3] - m_view_projection.r[3].m128_f32[1]
+		});
+
+		//Bottom plane
+
+		m_planes[3] = DirectX::XMPlaneNormalize({
+			m_view_projection.r[0].m128_f32[3] + m_view_projection.r[0].m128_f32[1],
+			m_view_projection.r[1].m128_f32[3] + m_view_projection.r[1].m128_f32[1],
+			m_view_projection.r[2].m128_f32[3] + m_view_projection.r[2].m128_f32[1],
+			m_view_projection.r[3].m128_f32[3] + m_view_projection.r[3].m128_f32[1]
+		});
+
+		//Near plane
+
+		m_planes[4] = DirectX::XMPlaneNormalize({
+			m_view_projection.r[0].m128_f32[2],
+			m_view_projection.r[1].m128_f32[2],
+			m_view_projection.r[2].m128_f32[2],
+			m_view_projection.r[3].m128_f32[2]
+		});
+
+		//Far plane
+
+		m_planes[5] = DirectX::XMPlaneNormalize({
+			m_view_projection.r[0].m128_f32[3] - m_view_projection.r[0].m128_f32[2],
+			m_view_projection.r[1].m128_f32[3] - m_view_projection.r[1].m128_f32[2],
+			m_view_projection.r[2].m128_f32[3] - m_view_projection.r[2].m128_f32[2],
+			m_view_projection.r[3].m128_f32[3] - m_view_projection.r[3].m128_f32[2]
+		});
+
+	}
+
+	bool CameraNode::InView(std::shared_ptr<MeshNode>& node)
```

Comment:
Comment in the wrong place but, could you add a bool in the settings to turn on and off camera culling? Thanks